article/app/controllers/ArticleController.scala

```diff
@@ -1,40 +1,174 @@
 
-class ArticleController(contentApiClient: ContentApiClient, val controllerComponents: ControllerComponents, ws: WSClient)(implicit context: ApplicationContext) extends BaseController
-    with RendersItemResponse with Logging with ImplicitControllerExecutionContext {
+class ArticleController(
+  contentApiClient: ContentApiClient,
+  val controllerComponents: ControllerComponents,
+  ws: WSClient
+)(implicit context: ApplicationContext)
+  extends BaseController
+    with RendersItemResponse
+    with Logging
+    with ImplicitControllerExecutionContext {
+
+  override def canRender(i: ItemResponse): Boolean = i.content.exists(ArticleController.isSupported)
+  override def renderItem(path: String)(implicit request: RequestHeader): Future[Result] = render(path, Some(Canonical))(buildResponse)
+
+  def renderLiveBlogJson(
+    path: String,
+    lastUpdate: Option[String],
+    rendered: Option[Boolean],
+    isLivePage: Option[Boolean]
+  ): Action[AnyContent] = {
+
+    Action.async { implicit request =>
+      val range = lastUpdate.map(ParseBlockId.fromBlockId) match {
+        case Some(ParsedBlockId(id)) => Some(SinceBlockId(id))
+        case Some(InvalidFormat) => None
+        case None => Some(Canonical)
+      }
+
+      range match {
+        case Some(r) =>
+          render(path, range) { page =>
+            r match {
+              case since: SinceBlockId => renderLiveBlogUpdates(page, since, isLivePage)
+              case render if rendered.contains(true) => buildResponse(page)
+              case block => blockText(page, 6)
+
+            }
+          }
+        case None =>
+          // indicates invalid block ID
+          Future.successful(Cached(10)(WithoutRevalidationResult(NotFound)))
+      }
+    }
+  }
+
+  def renderLiveBlog(path: String, page: Option[String], format: Option[String]): Action[AnyContent] = {
+    Action.async { implicit request =>
+      val range = page.map(ParseBlockId.fromPageParam) match {
+        case Some(ParsedBlockId(id)) => Some(PageWithBlock(id))
+        case Some(InvalidFormat) => None
+        case None => Some(Canonical)
+      }
 
-  private def isSupported(c: ApiContent) = c.isArticle || c.isLiveBlog || c.isSudoku
-  override def canRender(i: ItemResponse): Boolean = i.content.exists(isSupported)
-  override def renderItem(path: String)(implicit request: RequestHeader): Future[Result] = mapModel(path, Some(Canonical))(render(path, _))
+      range match {
+        case _ if request.isEmail =>
+          render(path, None)(buildResponse)
+        case Some(r) =>
+          render(path, range)(buildResponse)
+        case None =>
+          // indicates invalid block ID
+          Future.successful(Cached(10)(WithoutRevalidationResult(NotFound)))
+      }
+    }
+  }
 
+  def renderArticle(path: String): Action[AnyContent] = {
+    Action.async { implicit request =>
+      val range = if (request.isEmail || request.isGuuiJson) {
+        Some(ArticleBlocks)
+      } else None
```

Comment:
This block is kinda unreadable - why does something being an email/guuijson affect the range?

Reply:
It's a good question! Probably should add some comments.